### PR TITLE
DDC staking for CDN and storage roles

### PR DESCRIPTION
**Pull request by khssnv**
This PR delivers the following update.

1. Minimal staking lifecycle set of calls for Cere Network CDN and storage network participants,
2. New `spec_version` format with two minor digits related to `Dev` environment only updates.

### CODE REVIEW

File: frame/cere-staking/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+pub use pallet::*;
+
+const CERE_STAKING_ID: LockIdentifier = *b"cnmstake"; // Cere Network Maintainer's stake
+
+/// The balance type of this pallet.
+pub type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+parameter_types! {
+	pub MaxUnlockingChunks: u32 = 32;
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Why `32`?

File: frame/cere-staking/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+	}
+
+	/// Map from all locked "stash" accounts to the controller account.
+	#[pallet::storage]
+	#[pallet::getter(fn bonded)]
+	pub type Bonded<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, T::AccountId>;
+
+	/// The minimum active bond to become and maintain the role of a storage network participant.
+	#[pallet::storage]
+	pub type MinStorageBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Should it be readable?

File: frame/cere-staking/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+
+	/// The minimum active bond to become and maintain the role of a CDN participant.
+	#[pallet::storage]
+	pub type MinEdgeBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Should it be readable?

File: frame/cere-staking/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+			if <Ledger<T>>::contains_key(&controller) {
+				Err(Error::<T>::AlreadyPaired)?
+			}
+
+			// Reject a bond which is considered to be _dust_.
+			if value < T::Currency::minimum_balance() {
+				Err(Error::<T>::InsufficientBond)?
+			}
+
+			frame_system::Pallet::<T>::inc_consumers(&stash).map_err(|_| Error::<T>::BadState)?;
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: What is `inc_consumers` responsible for?

File: frame/cere-staking/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+	///
+	/// This is the latest planned era, depending on how the Session pallet queues the validator
+	/// set, it might be active or not.
+	#[pallet::storage]
+	#[pallet::getter(fn current_era)]
+	pub type CurrentEra<T> = StorageValue<_, EraIndex>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Let's have events on all state changes. For instance, we need to add event when `serve` and `store` executed.

File: frame/cere-staking/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+		#[pallet::weight(100_000)]
+		pub fn store(origin: OriginFor<T>, prefs: StoragePrefs) -> DispatchResult {
+			let controller = ensure_signed(origin)?;
+
+			let ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
+
+			ensure!(ledger.active >= MinStorageBond::<T>::get(), Error::<T>::InsufficientBond);
+			let stash = &ledger.stash;
+
+			// Can't participate in storate network if already participating in CDN.
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Fix typo in `storaTe`

File: frame/cere-staking/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+		#[pallet::weight(100_000)]
+		pub fn serve(origin: OriginFor<T>, prefs: EdgePrefs) -> DispatchResult {
+			let controller = ensure_signed(origin)?;
+
+			let ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
+
+			ensure!(ledger.active >= MinEdgeBond::<T>::get(), Error::<T>::InsufficientBond);
+			let stash = &ledger.stash;
+
+			// Can't participate in CDN if already participating in storate network.
```
Comment (AndreiNavoichyk) [inline frame/cere-staking/src/lib.rs]: Fix typo in `storaTe`

File: frame/cere-staking/src/lib.rs (hunk 2 above)
Comment (khssnv) [reply frame/cere-staking/src/lib.rs]: Sorry, it was my mistake during the review call. Storage items have default querying implementation and getters only affect how they are accessible from within the rust code. A getter is not required here.

https://docs.substrate.io/build/runtime-storage#getter-methods

File: frame/cere-staking/src/lib.rs (hunk 3 above)
Comment (khssnv) [reply frame/cere-staking/src/lib.rs]: Please check https://github.com/Cerebellum-Network/pos-network-node/pull/191#discussion_r1113889529

File: frame/cere-staking/src/lib.rs (hunk 4 above)
Comment (khssnv) [reply frame/cere-staking/src/lib.rs]: It increments a reference counter related to the given account to ensure another pallet does not delete it until our pallet uses it.

> The consumers reference counter indicating the number of other modules that currently depend on this account's existence.

More: https://docs.substrate.io/reference/account-data-structures#account-reference-counters

File: frame/cere-staking/src/lib.rs (hunk 1 above)
Comment (khssnv) [reply frame/cere-staking/src/lib.rs]: It is the [same in the FRAME staking pallet](https://github.com/Cerebellum-Network/pos-network-node/blob/c026e5a11f3ced2fbaaca194c44e017b5e7e915a/frame/staking/src/lib.rs#L352) and will not introduce an unexpected behavior for the users familiar with regular Cere Network staking.

File: bin/node/runtime/src/lib.rs [hunk 8]
```diff
@@ -128,7 +129,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	// and set impl_version to 0. If only runtime
 	// implementation changes and behavior does not, then leave spec_version as
 	// is and increment impl_version.
-	spec_version: 302,
+	spec_version: 30200,
```
Comment (Raid5594) [inline bin/node/runtime/src/lib.rs:132]: why spec version has this format?

File: frame/ddc-staking/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+	impl<T: Config> Pallet<T> {
+		/// Take the origin account as a stash and lock up `value` of its balance. `controller` will
+		/// be the account that controls it.
+		///
+		/// `value` must be more than the `minimum_balance` specified by `T::Currency`.
+		///
+		/// The dispatch origin for this call must be _Signed_ by the stash account.
+		///
+		/// Emits `Bonded`.
+		#[pallet::weight(100_000)]
```
Comment (Raid5594) [inline frame/ddc-staking/src/lib.rs:218]: Why is the weight hardcoded for all methods?

File: frame/cere-staking/src/lib.rs (hunk 5 above)
Comment (khssnv) [reply frame/cere-staking/src/lib.rs]: The original staking pallet does not emit an event from `validate` or `nominate` calls. Let's preserve this behavior? Other significant state updates are covered by existing events.

https://github.com/Cerebellum-Network/pos-network-node/blob/c026e5a11f3ced2fbaaca194c44e017b5e7e915a/frame/staking/src/pallet/mod.rs#L994
https://github.com/Cerebellum-Network/pos-network-node/blob/c026e5a11f3ced2fbaaca194c44e017b5e7e915a/frame/staking/src/pallet/mod.rs#L1034

File: frame/ddc-staking/src/lib.rs (hunk 9 above)
Comment (khssnv) [reply frame/ddc-staking/src/lib.rs:218]: This is a quick solution for a PoC. Realistic estimates can be generated by a benchmark we may introduce together with tests.

File: frame/ddc-staking/src/lib.rs [hunk 10]
```diff
@@ -16,7 +16,7 @@ use sp_std::prelude::*;
 
 pub use pallet::*;
 
-const CERE_STAKING_ID: LockIdentifier = *b"cnmstake"; // Cere Network Maintainer's stake
+const DDC_STAKING_ID: LockIdentifier = *b"cnmstake"; // DDC maintainer's stake
```
Comment (AndreiNavoichyk) [inline frame/ddc-staking/src/lib.rs]: What about renaming `cnmstake` -> `ddcstake`?

Comment (khssnv) [reply frame/ddc-staking/src/lib.rs]: Done https://github.com/Cerebellum-Network/pos-network-node/pull/191/commits/bc69158e16ee2d0b924ffb1ed2a7a8fde1cdb168

**Review by AndreiNavoichyk — APPROVED** [review]

**Pull request merged**